### PR TITLE
Fix #515

### LINKED ISSUE #515: Polyline edge routed over node
In the attached graph there is a single unexpected edge and I'm wondering why it is routed over a node. 

polyline

Not sure if this is a mis-configuration, an algorithm effect or just a situation that can't be helped?! Also the example resisted my attempts to minimize it, i.e. every time an element is removed the effect is not visible anymore (didn't try all combinations though :-)

Any insights?

[Polyline_edge_routed_over_node.json.zip](https://github.com/eclipse/elk/files/4292695/Polyline_edge_routed_over_node.json.zip)

### PR DESCRIPTION
**Pull request by uruuru**

### CODE REVIEW

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java [hunk 1]
```diff
@@ -323,6 +323,23 @@ private void processNode(final LNode node, final double layerLeftXPos, final dou
         
         for (LPort port : node.getPorts()) {
             KVector absolutePortAnchor = port.getAbsoluteAnchor();
+            
+            if (node.getType() == NodeType.NORTH_SOUTH_PORT) {
+                // North/south ports require special handling (see also #515):
+                // The dummy node that represents the north/south port will be placed at the very left x-coordinate
+                // of the layer currently processed. However, the actual x-coordinate of the port itself is very likely
+                // at a larger x-coordinate.
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java:331]: Might also be placed at the layer's right border if there are only outgoing edges, right?

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java [hunk 2]
```diff
@@ -323,6 +323,23 @@ private void processNode(final LNode node, final double layerLeftXPos, final dou
...
+                // The code below introduces bendpoints that prevent edge/node overlaps based on a condition that
+                // checks that a node's position is "far enough away" from the left-most x-coordinate of the layer.
+                // Consequently, we have to use the north/south port's x-coordinate instead of the dummy node's one.
+                LPort correspondingPort = (LPort) port.getProperty(InternalProperties.ORIGIN);
+                absolutePortAnchor.x = correspondingPort.getAbsoluteAnchor().x;
+                // While it makes sense either way, it is important to move the dummy node to the correct location
+                // as well. Otherwise an the #addBendPoint(...) method won't add the computed bendpoint as it
```
Comment (le-cds) [inline plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java:338]: "an the"

File: plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java (hunk 1 above)
Comment (uruuru) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java:331]: Good point. I didn't check that. The fix should still be the same, I guess?
I'll add a test.

Comment (uruuru) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java]: Behavior as expected: dummy placed at the very right of the layer for an outgoing edge - fix is the same. 

Note that the issue does not occur for north/south ports with incoming _and_ outgoing edges when the dummy is placed somewhere inside the layer. I added a testfile for this case as well, however.

Comment (le-cds) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java]: Those dummy nodes are placed just like regular nodes: their position inside the layer depends on the number of connections to each side, which sort of "pull" the node close to this or that border of the layer. The fix will be the same in all three conditions.

Comment (uruuru) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java]: I was trying to say: If there is one incoming and one outgoing edge, there's no need for a fix, as it's working properly already :).

Comment (le-cds) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java]: That depends, doesn't it? Perhaps the port is placed in a way that it ends up at the layer's left boundary. Thus, no additional bend point might be required for edges from the left.

Comment (uruuru) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java]: That's why I added "when the dummy is placed somewhere inside the layer" to my sentence (in my first answer 🙄).

Anyway, anything preventing this from merging?

Comment (le-cds) [reply plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/PolylineEdgeRouter.java]: Nope, I just wanted the comment to not confuse anyone who might be trying to debug any of this later on. 🙂

**Pull request merged**